File: app/src/main/java/com/yinduowang/installment/app/utils/LoadingUtils.java
package com.yinduowang.installment.app.utils;

import android.app.Activity;
import android.os.Handler;
import android.os.Parcel;
import android.os.Parcelable;

import com.tencent.bugly.crashreport.CrashReport;
import com.yinduowang.installment.R;
import com.yinduowang.installment.app.widget.LoadingDialog;

public class LoadingUtils implements Parcelable {

    private static final String TAG_LOADING_UTILS = "request_loading";
    // 延时时间
    private static final int DELAY_TIME = 100;

    private LoadingDialog dialog;

    private Runnable runnable;
    private Handler handler;
    private boolean isDismissing;
    private Activity activity;

    public static void init(Activity activity) {
        try {
            LoadingUtils loadingUtils = new LoadingUtils(activity);
            activity.getIntent().putExtra(TAG_LOADING_UTILS, loadingUtils);
        } catch (Exception e) {
            e.printStackTrace();
            CrashReport.postCatchedException(e);
        }
    }

    public static void show(Activity activity) {
        try {
            LoadingUtils loadingUtils = activity.getIntent().getParcelableExtra(TAG_LOADING_UTILS);
            if (activity != null && loadingUtils != null) {
                if (loadingUtils.isDismissing()) {
                    removeCallBacks(loadingUtils);
                    if (!loadingUtils.isShow()) {
                        loadingUtils.show();
                    }
                } else {
                    loadingUtils.show();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            CrashReport.postCatchedException(e);
        }
    }

    private static void removeCallBacks(LoadingUtils loadingUtils) {
        loadingUtils.getHandler().removeCallbacks(loadingUtils.getRunnable());
    }

    public static void dismiss(Activity activity) {
        try {
            LoadingUtils loadingUtils = activity.getIntent().getParcelableExtra(TAG_LOADING_UTILS);
            loadingUtils.dismiss();
        } catch (Exception e) {
            e.printStackTrace();
            CrashReport.postCatchedException(e);
        }
    }

    /**
     * 立即隐藏
     */
    public static void dismissNow(Activity activity) {
        try {
            LoadingUtils loadingUtils = activity.getIntent().getParcelableExtra(TAG_LOADING_UTILS);
            loadingUtils.dismissNow();
            removeCallBacks(loadingUtils);
        } catch (Exception e) {
            e.printStackTrace();
            CrashReport.postCatchedException(e);
        }
    }

    /**
     * 是否显示
     */
    public static boolean isShow(Activity activity) {
        try {
            LoadingUtils loadingUtils = activity.getIntent().getParcelableExtra(TAG_LOADING_UTILS);
            return loadingUtils.isShow();
        } catch (Exception e) {
            e.printStackTrace();
            CrashReport.postCatchedException(e);
            return false;
        }
    }

    private LoadingUtils(Activity activity) {
        this.activity = activity;
        dialog = new LoadingDialog(activity, R.style.dialog_waiting);

        this.activity.getIntent().putExtra(LoadingUtils.TAG_LOADING_UTILS, this);

        runnable = () -> {
            isDismissing = false;
            dismissNow();
        };
        handler = new Handler();
    }

    private void show() {
        if (this.activity != null && dialog != null && !dialog.isShowing()) {
            this.dialog.show();
        }
    }

    private boolean isShow() {
        if (this.activity != null && dialog != null) {
            return this.dialog.isShowing();
        }
        return false;
    }

    private void dismiss() {
        isDismissing = true;
        handler.postDelayed(runnable, DELAY_TIME);
    }

    private void dismissNow() {
        if (this.activity != null && dialog != null) {
            dialog.dismiss();
        }
    }

    protected LoadingUtils(Parcel in) {
    }

    public static final Creator<LoadingUtils> CREATOR = new Creator<LoadingUtils>() {
        @Override
        public LoadingUtils createFromParcel(Parcel in) {
            return new LoadingUtils(in);
        }

        @Override
        public LoadingUtils[] newArray(int size) {
            return new LoadingUtils[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
    }

    public Runnable getRunnable() {
        return runnable;
    }

    public Handler getHandler() {
        return handler;
    }

    public boolean isDismissing() {
        return isDismissing;
    }
}

File: app/src/main/java/com/yinduowang/installment/app/widget/LoadingUtils.kt
package com.yinduowang.installment.app.widget

import android.app.Activity
import android.os.Handler
import android.os.Parcel
import android.os.Parcelable
import com.tencent.bugly.crashreport.CrashReport
import com.yinduowang.installment.R
import com.yinduowang.installment.app.constant.Constant
import com.yinduowang.installment.app.widget.LoadingDialog

class LoadingUtils() : Parcelable {
    var dialog: LoadingDialog? = null
    var runnable: Runnable? = null
    var handler: Handler? = null
    var isDismissing: Boolean = false
    var activity:Activity?=null


    constructor(activity: Activity) :this()  {
        this.activity = activity
        dialog = LoadingDialog(activity, R.style.dialog_waiting)
        activity.intent.putExtra(Constant.TAG_LOADING_UTILS, this)
        runnable = Runnable {
            isDismissing = false
            dismissNow()
        }
        handler = Handler()
    }

    fun init(activity: Activity) {
        try {
            val loadingUtils = LoadingUtils(activity)
            activity.intent.putExtra(Constant.TAG_LOADING_UTILS, loadingUtils)
        } catch (e: Exception) {
            e.printStackTrace()
            CrashReport.postCatchedException(e)
        }
    }

    fun show(activity: Activity?) {
        try {
            val loadingUtils = activity!!.intent.getParcelableExtra<LoadingUtils>(Constant.TAG_LOADING_UTILS)
            if (loadingUtils != null) {
                if (loadingUtils.isDismissing) {
                    removeCallBacks(loadingUtils)
                    if (!loadingUtils.isShow()) {
                        loadingUtils.show()
                    }
                } else {
                    loadingUtils.show()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            CrashReport.postCatchedException(e)
        }
    }

    private fun removeCallBacks(loadingUtils: LoadingUtils) {
        loadingUtils.handler?.removeCallbacks(loadingUtils.runnable)
    }

    fun dismiss(activity: Activity) {
        try {
            val loadingUtils = activity.intent.getParcelableExtra<LoadingUtils>(Constant.TAG_LOADING_UTILS)
            loadingUtils.dismiss()
        } catch (e: Exception) {
            e.printStackTrace()
            CrashReport.postCatchedException(e)
        }
    }

    /**
     * 立即隐藏
     */
    fun dismissNow(activity: Activity) {
        try {
            val loadingUtils = activity.intent.getParcelableExtra<LoadingUtils>(Constant.TAG_LOADING_UTILS)
            loadingUtils.dismissNow()
            removeCallBacks(loadingUtils)
        } catch (e: Exception) {
            e.printStackTrace()
            CrashReport.postCatchedException(e)
        }
    }

    /**
     * 是否显示
     */
    fun isShow(activity: Activity): Boolean {
        return try {
            val loadingUtils = activity.intent.getParcelableExtra<LoadingUtils>(Constant.TAG_LOADING_UTILS)
            loadingUtils.isShow()
        } catch (e: Exception) {
            e.printStackTrace()
            CrashReport.postCatchedException(e)
            false
        }
    }

    private fun show() {
        if (dialog != null && !dialog?.isShowing!!) {
            dialog?.show()
        }
    }

    private fun isShow(): Boolean {
        return if (dialog != null) {
            dialog?.isShowing!!
        } else false
    }

    private fun dismiss() {
        isDismissing = true
        handler?.postDelayed(runnable, Constant.DELAY_TIME.toLong())
    }

    private fun dismissNow() {
        if (dialog != null) {
            dialog?.dismiss()
        }
    }

    constructor(parcel: Parcel) :this () {}

    override fun writeToParcel(dest: Parcel?, flags: Int) {
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<LoadingUtils> {
        override fun createFromParcel(parcel: Parcel): LoadingUtils {
            return LoadingUtils(parcel)
        }

        override fun newArray(size: Int): Array<LoadingUtils?> {
            return arrayOfNulls(size)
        }
    }

}